src/GenericCyclotomics.jl

```diff
@@ -0,0 +1,489 @@
+end
+
+# Basic manipulation
+
+zero(R::GenericCycloRing) = R()
+
+one(R::GenericCycloRing) = R(1)
+
+iszero(x::GenericCyclo) = isempty(x.f)
+
+isone(x::GenericCyclo) = isone(length(x.f)) && isone(get(x.f, base_ring(parent(x))(0)//base_ring(parent(x))(1), 0))  # TODO
+
+function is_unit(x::GenericCyclo) # TODO There could be more units than this.
+	if isone(length(x.f))
+		c=collect(x.f)[1][2]
+		return is_unit(c)
+	end
+	return false
+end
+
+characteristic(R::GenericCycloRing) = characteristic(base_ring(R))
+
+# Canonical unit
+
+canonical_unit(x::GenericCyclo) = parent(x)(1)  # TODO
+
+# String I/O
+
+function show(io::IO, ::MIME"text/plain", R::GenericCycloRing)
+	io = pretty(io)
+	println(io, "Generic cyclotomic ring", Indent())
+	println(io, "over ", base_ring(base_ring(R)))
+	print(io, "dependent on ", join(gens(base_ring(R)), ", "))
+	if R.congruence !== nothing
+		print(io, "\nwhere ", gens(base_ring(R))[1], " is congruent to ", R.congruence[1], " modulo ", R.congruence[2])
+	end
+end
+
+function show(io::IO, R::GenericCycloRing)
+	print(io, "Generic cyclotomic ring")
+	if !is_terse(io)
+		print(io, " over ", base_ring(base_ring(R)))
+	end
+end
+
+function show(io::IO, x::GenericCyclo)  # TODO
+	if iszero(x)
+		print(io, "0")
+		return
+	end
+	for (i, (argument, modulus)) in enumerate(x.f)
+		if iszero(argument)
+			print(io, modulus)
+		elseif isone(modulus)
+			print(io, "exp(2π𝑖($(argument)))")
+		elseif is_monomial(modulus)
+			print(io, "$(modulus)*exp(2π𝑖($(argument)))")
+		else
+			print(io, "($(modulus))*exp(2π𝑖($(argument)))")
+		end
+		if i < length(x.f)
+			print(io, " + ")
+		end
+	end
+end
+
+# Unary operations
+
+function -(x::GenericCyclo)
+	f=deepcopy(x.f)
+	for (key,value) in f
+		f[key]=-value
+	end
+	return parent(x)(f, simplify=false)
+end
+
+function conj(x::GenericCyclo)
+	f=Dict{UPolyFrac, UPoly}()
+	for (key,value) in x.f
+		f[-key]=value
+	end
+	return parent(x)(f)
+end
+
+# Binary operations
+
+function +(x::GenericCyclo, y::GenericCyclo)
+	parent(x) != parent(y) && error("Incompatible rings")
+	f=mergewith(+, x.f, y.f)
+	strip_zeros!(f)
+	return parent(x)(f)
+end
+
+-(x::GenericCyclo, y::GenericCyclo) = x+(-y)  # TODO make this more efficient?
+
+function *(x::GenericCyclo, y::GenericCyclo)
+	parent(x) != parent(y) && error("Incompatible rings")
+	f = Dict{UPolyFrac, UPoly}()
+	for (key1,value1) in x.f
+		for (key2,value2) in y.f
+			key=key1+key2
+			if haskey(f, key)
+				f[key]+=value1*value2
+			else
+				f[key]=value1*value2
+			end
+		end
+	end
+	return parent(x)(f)
+end
+
++(x::UPoly, y::GenericCyclo) = parent(y)(x)+y
+
++(x::GenericCyclo, y::UPoly) = y+x
+
+-(x::UPoly, y::GenericCyclo) = parent(y)(x)-y
+
+-(x::GenericCyclo, y::UPoly) = x-parent(x)(y)
+
+*(x::UPoly, y::GenericCyclo) = parent(y)(x)*y
+
+*(x::GenericCyclo, y::UPoly) = y*x
+
+# Comparison
+
+function ==(x::GenericCyclo, y::GenericCyclo)
+	parent(x) != parent(y) && error("Incompatible rings")
+	return iszero(x-y)
```

Comment:
Yes, unfortunately there are different representations for the same generic cyclotomic. This can be improved by using Zumbroich bases for the constant part of the exponents. But I only found out about them relatively recently. I'm not sure if they completely fix the problem though. Maybe there are some nasty cases...